### PR TITLE
[jax2tf] First draft of converting sort_p using TF2XLA

### PR DESCRIPTION
**Pull request by bchetioui**

### CODE REVIEW

File: jax/experimental/jax2tf/jax2tf.py [hunk 1]
```diff
@@ -1098,6 +1098,30 @@ def func1(*jax_args):
 
 tf_impl[lax.scan_p] = _scan
 
+def _sort(*operand: tf.Tensor, dimension: int = -1,
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:1101]: Shouldn't operand be of type `TfVal`, which includes `tf.Tensor` but also numpy arrays.

File: jax/experimental/jax2tf/jax2tf.py [hunk 2]
```diff
@@ -1098,6 +1098,30 @@ def func1(*jax_args):
...
+          is_stable: bool = True) -> Tuple[TfVal, ...]:
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:1102]: We prefer to not use default values, we want errors if the argument is missing. I think that the `sort_p` primitive always has the `dimension` and `is_stable`. I propose
```
def _sort(*operand: TfVal, *, dimension: int, is_stable: bool)
```

File: jax/experimental/jax2tf/jax2tf.py [hunk 3]
```diff
@@ -1098,6 +1098,30 @@ def func1(*jax_args):
...
+  if len(operand) == 0:
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:1103]: This is probably more like an internal assertions, because we should never have a `sort_p` primitive with no arguments (if we do, the `TypeError` should be in `lax.sort`)

Comment (gnecula) [reply jax/experimental/jax2tf/jax2tf.py:1103]: However, if you do want to keep it also for extra safety, I think it is fine.

File: jax/experimental/jax2tf/jax2tf.py [hunk 4]
```diff
@@ -1098,6 +1098,30 @@ def func1(*jax_args):
...
+    raise TypeError("Sort requires at least one operand")
+  if len(operand) > 1:
+    raise NotImplementedError("TODO: multiple tensors")
+  dimension = lax._canonicalize_axis(dimension, len(operand[0].shape))
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:1107]: I believe that we do not need `_canonicalize_axis` because we can assume here that once `lax.sort` has been turned into a `sort_p.bind` all arguments have been sanitized. If you look at `_sort_translation_rule` in `lax.py` you will see that it also does not do any argument sanitization.

File: jax/experimental/jax2tf/jax2tf.py [hunk 5]
```diff
@@ -1098,6 +1098,30 @@ def func1(*jax_args):
...
+  if dimension == lax._canonicalize_axis(-1, len(operand[0].shape)):
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:1108]: Perhaps we should raise `NotImplementedError` if we see `is_stable`

File: jax/experimental/jax2tf/jax2tf.py [hunk 6]
```diff
@@ -1098,6 +1098,30 @@ def func1(*jax_args):
...
+    # TODO: rewrite XlaSort to allow stable sorting.
+    return tfxla.sort(operand)
+  else:
+    # tfxla.sort only supports sorting the last axis of its operands.
+    # TODO: fix XlaSort to allow an axis parameter. In the meantime, we transpose the
+    # tensors before and after sorting if it does not happen on the last axis.
+    shape_range = list(range(len(operand[0].shape)))
+    permutation     = shape_range[ dimension + 1:] + shape_range[: dimension + 1]
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:1116]: Minor: Google code style would not like the extra spaces (before = or after [ or after : )

File: jax/experimental/jax2tf/jax2tf.py [hunk 7]
```diff
@@ -1098,6 +1098,30 @@ def func1(*jax_args):
...
+    inv_permutation = shape_range[-dimension - 1:] + shape_range[:-dimension - 1]
```
Comment (gnecula) [inline jax/experimental/jax2tf/jax2tf.py:1117]: It is somewhat dangerous to add transpositions because they may be expensive. Eventually, we will see if we can fix XlaSort to specify the axis. Meanwhile, maybe we can just raise UnimplementedError. Or, if we do add transposition, there are multiple ways of doing it, e.g, just swap the `dimension` with the last axis.

File: jax/experimental/jax2tf/tests/primitives_test.py [hunk 8]
```diff
@@ -167,6 +167,23 @@ def test_pad(self, harness: primitive_harness.Harness):
     self.ConvertAndCompare(harness.dyn_fun, *harness.dyn_args_maker(self.rng()),
                            with_function=False)
 
+  @primitive_harness.parameterized(primitive_harness.lax_sort)
+  def test_sort(self, harness: primitive_harness.Harness):
+    if harness.params["dtype"] is dtypes.bfloat16:
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/primitives_test.py:172]: I would maybe merge the first two conditionals. WDYT?

File: jax/experimental/jax2tf/tests/primitives_test.py [hunk 9]
```diff
@@ -167,6 +167,23 @@ def test_pad(self, harness: primitive_harness.Harness):
...
+      # TODO: implement bfloat16 in XlaSort
+      raise unittest.SkipTest("bfloat16 not implemented")
+    if harness.params["dtype"] in jtu.dtypes.complex:
+      # TODO: implement complex support in XlaSort
+      raise unittest.SkipTest("complex not implemented")
+    if harness.params["is_stable"]:
+      # TODO: implement stable sort support in XlaSort
+      raise unittest.SkipTest("stable sort not implemented")
+    if len(harness.arg_descriptors) > 3:
+      # TODO: make default comparison operator in TF2XLA and lax.sort compatible
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/primitives_test.py:182]: We should talk about this one

File: jax/experimental/jax2tf/tests/primitive_harness.py [hunk 10]
```diff
@@ -186,6 +186,46 @@ def parameterized(harness_group: Iterable[Harness],
   ]
 )
 
+
+lax_sort = jtu.cases_from_list( # one array, random data, all axes, all dtypes
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/primitive_harness.py:190]: I have just discovered while working on #3700 that jtu.cases_from_list may drop test cases if JAX_NUM_GENERATED_TESTS is too small. Let's use `tuple` instead

File: jax/experimental/jax2tf/tests/primitive_harness.py [hunk 11]
```diff
@@ -186,6 +186,46 @@ def parameterized(harness_group: Iterable[Harness],
...
+          shape=shape,
+          dimension=dimension,
+          dtype=dtype,
+          is_stable=is_stable)
+  for dtype in jtu.dtypes.all
+  for shape in [(5,), (5, 7)]
+  for dimension in range(len(shape))
+  for is_stable in [False, True]
+) + jtu.cases_from_list( # one array, potential edge cases
+  Harness(f"one_array_shape={jtu.format_shape_dtype_string(arr.shape, arr.dtype)}_axis={dimension}_isstable={is_stable}",
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/primitive_harness.py:203]: The names of the harnesses have to be unique. This works here because the `arr.shape` is different from what you have used already. I would make it even safer (and easier to find) by changing the `one_array_` prefix to `one_special_array`

**Comment by gnecula** [comment]
I think that the test failure is because your branch is not sync-ed to the main HEAD, where there are changes for sort_p

**Comment by gnecula** [comment]
LGTM

**Review by gnecula — APPROVED** [review]
LGTM. 

I have verified that the internal tests pass.
